src/main/scala/beam/router/BeamSkimmer.scala

```diff
@@ -41,50 +63,175 @@ class BeamSkimmer(services: Option[BeamServices] = None) {
+            skim
+          case None =>
+            Skim(1.1, 0, 0.6, 0)
+        }
+    }
+    (pooled.time / solo.time, pooled.cost / solo.cost)
+  }
+
+  private def distanceAndTime(mode: BeamMode, origin: Location, destination: Location) = {
     val speed = mode match {
-      case BeamMode.CAR     => carSpeedMeterPerSec
-      case BeamMode.TRANSIT => transitSpeedMeterPerSec
-      case BeamMode.BIKE    => bicycleSpeedMeterPerSec
-      case _                => walkSpeedMeterPerSec
+      case CAR | CAV | RIDE_HAIL                                      => carSpeedMeterPerSec
+      case RIDE_HAIL_POOLED                                           => carSpeedMeterPerSec * 1.1
+      case TRANSIT | WALK_TRANSIT | DRIVE_TRANSIT | RIDE_HAIL_TRANSIT => transitSpeedMeterPerSec
+      case BIKE                                                       => bicycleSpeedMeterPerSec
+      case _                                                          => walkSpeedMeterPerSec
     }
     val travelDistance: Int = Math.ceil(GeoUtils.minkowskiDistFormula(origin, destination)).toInt
     val travelTime: Int = Math
       .ceil(travelDistance / speed)
       .toInt + ((travelDistance / trafficSignalSpacing).toInt * waitingTimeAtAnIntersection).toInt
-    val travelCost: Double = services
-      .map(
-        x =>
-          DrivingCost.estimateDrivingCost(
-            new BeamLeg(departureTime, mode, travelTime, new BeamPath(null, null, None, null, null, travelDistance)),
-            vehicleTypeId,
-            x
+    (travelDistance, travelTime)
+  }
+
+  private def getSkimValue(time: Int, mode: BeamMode, orig: Id[TAZ], dest: Id[TAZ]): Option[Skim] = {
+    skims.get((timeToBin(time), mode, orig, dest)) match {
+      case Some(skim) =>
+        Some(skim)
+      case None =>
+        previousSkims.get((timeToBin(time), mode, orig, dest))
+    }
+  }
+
+  def observeTrip(trip: EmbodiedBeamTrip, beamServices: BeamServices) = {
+    val mode = trip.tripClassifier
+    val correctedTrip = mode match {
+      case WALK =>
+        trip.beamLegs()
+      case RIDE_HAIL =>
+        trip.beamLegs().drop(1).dropRight(1)
+      case _ =>
+        trip.beamLegs().drop(1).dropRight(1)
+    }
+    val origLeg = correctedTrip.head
+    val origCoord = beamServices.geo.wgs2Utm(origLeg.travelPath.startPoint.loc)
+    val origTaz = beamServices.tazTreeMap
+      .getTAZ(origCoord.getX, origCoord.getY)
+      .tazId
+    val destLeg = correctedTrip.last
+    val destCoord = beamServices.geo.wgs2Utm(destLeg.travelPath.endPoint.loc)
+    val destTaz = beamServices.tazTreeMap
+      .getTAZ(destCoord.getX, destCoord.getY)
+      .tazId
+    val timeBin = timeToBin(origLeg.startTime)
+    val key = (timeBin, mode, origTaz, destTaz)
+    val payload =
+      Skim(trip.totalTravelTimeInSecs.toDouble, trip.beamLegs().map(_.travelPath.distanceInM).sum, trip.costEstimate, 1)
+    skims.get(key) match {
+      case Some(existingSkim) =>
+        val newPayload = Skim(
```

Comment:
the current implementation seems to keep average over all iterations - no memory limit to reflect current state faster, e.g. only last 2-5 iterations? due to the way that averaging works and how the simulation system evolves, the skim data might be representing "iteration-i/2" state instead of more recent one. just to make sure, this is what is needed for the applications using the skim. e.g. the ridehail repositioning history collector just keeps last 2 iteration history (and not average over all seen iterations till now)